nl_server/config_reader.py

```diff
@@ -0,0 +1,214 @@
+from shared.lib import gcs
+
+# catalog paths
+_CATALOG_CODE_PATH = Path(__file__).parent / '../deploy/nl/catalog.yaml'
+_CATALOG_MOUNT_PATH = '/datacommons/nl/catalog.yaml'
+_CATALOG_USER_PATH_SUFFIX = 'datacommons/nl/custom_catalog.yaml'
+_CATALOG_TMP_PATH = '/tmp/catalog.yaml'
+
+# env paths
+_ENV_CODE_PATH = (Path(__file__).parent /
+                  '../deploy/helm_charts/envs/autopush.yaml')
+_ENV_MOUNT_PATH = '/datacommons/nl/env.yaml'
+_ENV_USER_PATH = Path(__file__).parent / 'custom_dc_env.yaml'
+
+_SUPPORTED_VERSIONS = [1]
+
+
+def _merge_dicts(x, y):
+  res = {}
+  for k, v in x.items():
+    if v:
+      res[k] = v
+  for k, v in y.items():
+    if v:
+      res[k] = v
+  return res
+
+
+def read_catalog() -> Catalog:
+  """
+  Reads the catalog from the config files and merges them together.
+
+  One Config file could exist in several places depending on the environment.
+
+  - `../../deploy/nl/catalog.yaml`
+  - `/datacommons/nl/catalog.yaml`
+  - `${USER_DATA_PATH}/datacommons/nl/custom_catalog.yaml`
+
+  Note here ${USER_DATA_PATH} could be a local path or a GCS path (gcs://)
+  """
+  # Note the check order here is (somewhat) important. The later config could
+  # overwrite the earlier config.
+
+  all_paths = []
+  # Check the codebase config file
+  if os.path.exists(_CATALOG_CODE_PATH):
+    all_paths.append(_CATALOG_CODE_PATH)
+
+  # Check mounted config file
+  if os.path.exists(_CATALOG_MOUNT_PATH):
+    all_paths.append(_CATALOG_MOUNT_PATH)
+
+  # Then check user set config file
+  # TODO: make this generic and independent of custom DC.
+  # Currently this depends on the assumption of custom DC user sub path.
+  user_data_path = custom_dc_util.get_custom_dc_user_data_path()
+
+  if gcs.is_gcs_path(user_data_path):
+    full_gcs_path = os.path.join(user_data_path, _CATALOG_USER_PATH_SUFFIX)
+    if gcs.download_blob_by_path(full_gcs_path, _CATALOG_TMP_PATH):
+      all_paths.append(_CATALOG_TMP_PATH)
+  else:
+    full_user_path = os.path.join(user_data_path, _CATALOG_USER_PATH_SUFFIX)
+    if os.path.exists(full_user_path):
+      all_paths.append(full_user_path)
+
+  # Now load and merge all the catalog config files
+  catalog = Catalog(
+      version='',
+      indexes={},
+      models={},
+  )
+  for p in all_paths:
+    logging.info('Loading index and model catalog from: %s', p)
+    with open(p) as f:
+      partial_catalog = yaml.safe_load(f.read())
+      # read version
+      ver = catalog.version
+      this_ver = partial_catalog['version']
+      if (ver and this_ver and ver != this_ver):
+        raise ValueError(
+            f'Inconsistent version in config files: {ver} and {this_ver}')
+
+      if this_ver not in _SUPPORTED_VERSIONS:
+        raise ValueError(f'Unknown version: {this_ver}')
+      catalog.version = this_ver
+
+      indexes = catalog.indexes
+      models = catalog.models
+
+      # TODO: when another version is added, extract separate version read as
+      # a function.
```

Comment:
Consider already extracting as a function?

Reply:
I really think the yaml conversion should happen offline during import or system update, not at runtime..